Sources/OAuthenticator/Services/Mastodon.swift

```diff
@@ -0,0 +1,164 @@
+
+	public static func tokenHandling(with parameters: UserTokenParameters) -> TokenHandling {
+		TokenHandling(authorizationURLProvider: authorizationURLProvider(with: parameters),
+					  loginProvider: loginProvider(with: parameters),
+					  refreshProvider: refreshProvider(with: parameters))
+	}
+
+	static func authorizationURLProvider(with parameters: UserTokenParameters) -> TokenHandling.AuthorizationURLProvider {
+		return { credentials in
+			var urlBuilder = URLComponents()
+
+			urlBuilder.scheme = Mastodon.scheme
+			urlBuilder.host = parameters.host
+			urlBuilder.path = Mastodon.authorizePath
+			urlBuilder.queryItems = [
+				URLQueryItem(name: Mastodon.clientIDKey, value: credentials.clientId),
+				URLQueryItem(name: Mastodon.redirectURIKey, value: credentials.callbackURL.absoluteString),
+				URLQueryItem(name: Mastodon.responseTypeKey, value: Mastodon.responseTypeCode),
+				URLQueryItem(name: Mastodon.scopeKey, value: credentials.scopeString)
+			]
+
+			guard let url = urlBuilder.url else {
+				throw AuthenticatorError.missingAuthorizationURL
+			}
+
+			return url
+		}
+	}
+
+	static func authenticationRequest(with parameters: UserTokenParameters, url: URL, appCredentials: AppCredentials) throws -> URLRequest {
+		let code = try url.authorizationCode
+
+		var urlBuilder = URLComponents()
+
+		urlBuilder.scheme = Mastodon.scheme
+		urlBuilder.host = parameters.host
+		urlBuilder.path = Mastodon.tokenPath
+		urlBuilder.queryItems = [
+			URLQueryItem(name: Mastodon.grantTypeKey, value: Mastodon.grantTypeAuthorizationCode),
+			URLQueryItem(name: Mastodon.clientIDKey, value: appCredentials.clientId),
+			URLQueryItem(name: Mastodon.clientSecretKey, value: appCredentials.clientPassword),
+			URLQueryItem(name: Mastodon.redirectURIKey, value: appCredentials.callbackURL.absoluteString),
+			URLQueryItem(name: Mastodon.codeKey, value: code),
+			URLQueryItem(name: Mastodon.scopeKey, value: appCredentials.scopeString)
+		]
+
+		guard let url = urlBuilder.url else {
+			throw AuthenticatorError.missingTokenURL
+		}
+
+		var request = URLRequest(url: url)
+
+		request.httpMethod = "POST"
+		request.setValue("application/json", forHTTPHeaderField: "Accept")
+
+		return request
+	}
+
+	static func loginProvider(with parameters: UserTokenParameters) -> TokenHandling.LoginProvider {
+		return { url, appCredentials, tokenURL, urlLoader in
+			let request = try authenticationRequest(with: parameters, url: url, appCredentials: appCredentials)
+
+			let (data, _) = try await urlLoader(request)
+
+			let response = try JSONDecoder().decode(Mastodon.AppAuthResponse.self, from: data)
+
+			return response.login
+		}
+	}
+
+	static func refreshProvider(with parameters: UserTokenParameters) -> TokenHandling.RefreshProvider {
+		return { login, appCredentials, urlResponseProvider in
+			throw AuthenticatorError.refreshUnsupported
+		}
+	}
+
+	public static func register(with parameters: UserTokenParameters) async throws -> AppRegistrationResponse {
```

Comment:
What do you think about making this function take a `URLResponseProvider`? That would allow it to be testable without hitting the network. Maybe we could make `URLSession.defaultProvider` public to give it a default parameter? I think that could be an improvement for the `Authenticator` class too.

Reply:
Good point. I added a parameter `URLResponseProvider` which will indeed make the function testable.